src/python/pants/engine/target.py

```diff
@@ -610,23 +640,43 @@ def generate_subtarget(base_target: _Tgt, *, full_file_name: str) -> _Tgt:
     are able to deduce specifically which files are being used, we can use only the files we care
     about, rather than the entire `sources` field.
     """
+    if not base_target.has_field(Dependencies):
+        raise ValueError(
+            f"Target {base_target.address.spec} of type {type(base_target).__qualname__} does "
+            f"not support dependencies, and thus cannot depend on {full_file_name}."
+        )
+
     relativized_file_name = (
         PurePath(full_file_name).relative_to(base_target.address.spec_path).as_posix()
     )
 
-    base_target_field_values = {
-        field.alias: (
-            field.value
-            if isinstance(field, PrimitiveField)
-            else field.sanitized_raw_value  # type: ignore[attr-defined]
+    found_source_match = False
+    generated_target_fields = {}
+    for field in base_target.field_values.values():
+        if isinstance(field, Sources) and bool(
+            matches_filespec(field.filespec, paths=(full_file_name,))
+        ):
+            found_source_match = True
+            value = (relativized_file_name,)
+        else:
+            value = (
+                field.value
+                if isinstance(field, PrimitiveField)
+                else field.sanitized_raw_value  # type: ignore[attr-defined]
+            )
+        generated_target_fields[field.alias] = value
+
+    if not found_source_match:
+        # TODO: This should probably be ResolveError, but this function is in the wrong place
+        # for that.
+        sources_fields = [
+            field for field in base_target.field_values.values() if isinstance(field, Sources)
+        ]
+        sources_fields_hint = f" It only matches: {sources_fields}." if sources_fields else ""
```

Comment:
I'm not following this. Are you checking for multiple `Sources` field because it's technically possible for a class to multiple register?

This error message makes it sound like those `Sources` fields would match the file name, but we aren't calling `matches_filespec` to verify that.

Maybe simplify to a) assume there is only one `Sources` field, and b) move the check up higher when you first check `if isinstance(field, Sources)`, and error if the field is registered but the filespec doesn't match.

Reply:
> This error message makes it sound like those Sources fields would match the file name, but we aren't calling matches_filespec to verify that.

We are, a few lines up.